File: backend/lib/feedbackStore.js
const fs = require('fs');
const path = require('path');

class FeedbackStore {
  constructor(filePath){
    this.filePath = filePath;
    this._ready = this._init();
    this._writeQueue = Promise.resolve();
  }

  async _init(){
    const dir = path.dirname(this.filePath);
    await fs.promises.mkdir(dir, { recursive: true });
    try {
      await fs.promises.access(this.filePath, fs.constants.F_OK);
    } catch (err){
      await fs.promises.writeFile(this.filePath, '', { encoding: 'utf8' });
    }
  }

  async save(entry){
    await this._ready;
    const line = `${JSON.stringify(entry)}\n`;
    this._writeQueue = this._writeQueue.then(() => fs.promises.appendFile(this.filePath, line, { encoding: 'utf8' }));
    return this._writeQueue.catch(err => {
      // сбрасываем очередь, чтобы последующие вызовы могли выполниться
      this._writeQueue = Promise.resolve();
      throw err;
    });
  }
}

module.exports = FeedbackStore;

File: backend/feedback_store.py
"""Utilities for persisting feedback submissions."""

from __future__ import annotations

import asyncio
import json
from pathlib import Path
from typing import Any, Dict


class FeedbackStore:
    """Append-only JSONL store for feedback submissions."""

    def __init__(self, file_path: str | Path):
        self.file_path = Path(file_path)
        self._write_lock = asyncio.Lock()
        self._init_lock = asyncio.Lock()
        self._initialised = False

    async def _ensure_file(self) -> None:
        if self._initialised:
            return
        async with self._init_lock:
            if self._initialised:
                return
            self.file_path.parent.mkdir(parents=True, exist_ok=True)
            if not self.file_path.exists():
                self.file_path.write_text("", encoding="utf-8")
            self._initialised = True

    async def save(self, entry: Dict[str, Any]) -> None:
        """Persist a feedback entry as a JSON line."""
        await self._ensure_file()
        line = json.dumps(entry, ensure_ascii=False) + "\n"
        async with self._write_lock:
            loop = asyncio.get_running_loop()
            await loop.run_in_executor(None, self._append_line, line)

    def _append_line(self, line: str) -> None:
        with self.file_path.open("a", encoding="utf-8") as handle:
            handle.write(line)
